examples/experimental/imagenet_nas_arch_pytorch/model_def.py

```diff
@@ -11,25 +11,121 @@
+from torch.optim.lr_scheduler import _LRScheduler
 import torchvision.transforms as transforms
 from torch import nn
 
 import determined as det
 from data import ImageNetDataset
 from determined.pytorch import DataLoader, LRScheduler, PyTorchTrial, reset_parameters
 from model import NetworkImageNet
-from utils import AutoAugment, CrossEntropyLabelSmooth, Cutout, HSwish, Swish, accuracy
+from utils import (
+    RandAugment,
+    CrossEntropyLabelSmooth,
+    Cutout,
+    HSwish,
+    Swish,
+    accuracy,
+    AvgrageMeter,
+)
 
 Genotype = namedtuple("Genotype", "normal normal_concat reduce reduce_concat")
 
 activation_map = {"relu": nn.ReLU, "swish": Swish, "hswish": HSwish}
 
 
+class LinearLRScheduler(_LRScheduler):
+    def __init__(self, optimizer, max_epochs, warmup_epochs, last_epoch=-1):
+        self.optimizer = optimizer
+        self.warmup_epochs = warmup_epochs
+        self.max_epochs = max_epochs
+        self.last_epoch = last_epoch
+        super(LinearLRScheduler, self).__init__(optimizer, last_epoch)
+
+    def get_lr(self):
+        if self.last_epoch < self.warmup_epochs:
+            lr_mult = (self.last_epoch + 1) / self.warmup_epochs
+        elif self.max_epochs - self.last_epoch > self.warmup_epochs:
+            lr_mult = (self.max_epochs - self.warmup_epochs - self.last_epoch) / (
+                self.max_epochs - self.warmup_epochs
+            )
+        else:
+            lr_mult = (self.max_epochs - self.last_epoch) / (
+                (self.last_epoch - self.warmup_epochs) * 5
+            )
+        return [base_lr * lr_mult for base_lr in self.base_lrs]
+
+
+class EfficientNetScheduler(_LRScheduler):
+    def __init__(
+        self, optimizer, max_epochs, warmup_epochs, gamma, decay_every, last_epoch=-1
+    ):
+        self.optimizer = optimizer
+        self.warmup_epochs = warmup_epochs
+        self.max_epochs = max_epochs
+        self.last_epoch = last_epoch
+        self.gamma = gamma
+        self.decay_every = decay_every
+        super(EfficientNetScheduler, self).__init__(optimizer, last_epoch)
+
+    def get_lr(self):
+        if self.last_epoch < self.warmup_epochs:
+            lr_mult = (self.last_epoch + 1) / self.warmup_epochs
+        else:
+            lr_mult = self.gamma ** (int((self.last_epoch + 1) / self.decay_every))
+        return [base_lr * lr_mult for base_lr in self.base_lrs]
+
+
+class CosineWrapper(torch.optim.lr_scheduler.CosineAnnealingLR):
+    def __init__(self, optimizer, max_epochs, warmup_epochs, eta_min=0, last_epoch=-1):
+        self.warmup_epochs = warmup_epochs
+        super(CosineWrapper, self).__init__(optimizer, max_epochs, eta_min, last_epoch)
+
+    def get_lr(self):
+        if self.last_epoch < self.warmup_epochs:
+            return [
+                (self.last_epoch + 1) / self.warmup_epochs * b_lr
+                for b_lr in self.base_lrs
+            ]
+        return super(CosineWrapper, self).get_lr()
+
+    def _get_closed_form_lr(self):
```

Comment:
blocking: This looks doubled as get_lr. I believe we can just have get_lr since it's being treated as a get_closed_form_lr anyways.

Reply:
Lets then set a helper function that handles this inner if statement to prevent duplicate code.

Reply:
i created a lr warmup wrapper for any lr scheduler.